cpp/oneapi/dal/backend/primitives/reducer.cpp

```diff
@@ -0,0 +1,221 @@
+                            work_items_per_group);
+}
+
+template <typename Float, typename BinaryFunctor, typename UnaryFunctor, layout Layout>
+cl::sycl::event reducer_singlepass<Float, BinaryFunctor, UnaryFunctor, Layout>::operator()(
+    const Float* input,
+    Float* output,
+    std::int64_t vector_size,
+    std::int64_t n_vectors) {
+    return this->operator()(input, output, vector_size, n_vectors, this->max_work_group_size);
+}
+
+template <typename Float, typename BinaryFunctor, typename UnaryFunctor, layout Layout>
+cl::sycl::event reducer_singlepass<Float, BinaryFunctor, UnaryFunctor, Layout>::operator()(
+    array<Float> input,
+    array<Float> output,
+    std::int64_t vector_size,
+    std::int64_t n_vectors) {
+    return this->operator()(input, output, vector_size, n_vectors, this->max_work_group_size);
+}
+
+//Direct instantiation
+template struct reducer_singlepass<float,
+                                   binary_functor<float, binary_operation::sum>,
+                                   unary_functor<float, unary_operation::identity>,
+                                   layout::row_major>;
+template struct reducer_singlepass<double,
+                                   binary_functor<double, binary_operation::sum>,
+                                   unary_functor<double, unary_operation::identity>,
+                                   layout::row_major>;
+template struct reducer_singlepass<float,
+                                   binary_functor<float, binary_operation::sum>,
+                                   unary_functor<float, unary_operation::identity>,
+                                   layout::col_major>;
+template struct reducer_singlepass<double,
+                                   binary_functor<double, binary_operation::sum>,
+                                   unary_functor<double, unary_operation::identity>,
+                                   layout::col_major>;
+template struct reducer_singlepass<float,
+                                   binary_functor<float, binary_operation::sum>,
+                                   unary_functor<float, unary_operation::square>,
+                                   layout::row_major>;
+template struct reducer_singlepass<double,
+                                   binary_functor<double, binary_operation::sum>,
+                                   unary_functor<double, unary_operation::square>,
+                                   layout::row_major>;
+template struct reducer_singlepass<float,
+                                   binary_functor<float, binary_operation::sum>,
+                                   unary_functor<float, unary_operation::square>,
+                                   layout::col_major>;
+template struct reducer_singlepass<double,
+                                   binary_functor<double, binary_operation::sum>,
+                                   unary_functor<double, unary_operation::square>,
+                                   layout::col_major>;
```

Comment:
Why do you want to instantiate all of them explicitly and not to keep implementation in headers?